### PR TITLE
[BNTY] Mech Update

### PR DESCRIPTION
**Pull request by MLGTASTICa**
## About The Pull Request
This PR adds the following mech features
1)The pilot can be attacked/shot by any frontal attack as long as the hatch is not closed
2)Mech thrusters , which can also z-move up and down
3)Full auto mech weaponary
4)Adds the mech blade(Craftable with any material with hardness)
      -adds hardness to materials that were lacking it and should be useable in general for making blades.
5)Adds Mech Stomping (mechs will stomp roaches and lying down humans with varying damage based on type of legs/tread)

7)More mech ballistics
   Bobby shotgun - 12 rounds , 8 second reload , takes in small shotgun ammo packets and shell casings
   C35R smg - dependant on box loaded, 6 second reload ,high accuracy , 600 FULL auto, takes in SMG magazines or pistol ammo packets
  PK -  takes in only large rifle ammo boxes,  5 second reload , low accuracy , 400 FULL Auto
 Mags can be ejected from any of these modules by attacking in hand or attacking themselves in the mech UI.
8)Mining Mech QOL
    Ores that a mech walks over are now auto-inserted in any ore box the mech is carrying
    Mech plasmacutter now penetrates 5 tiles of minerals , instead of 1.
9)Added the mech energy shield (which works by draining the energy cell of the mech , currently at a rate of 20 per point of damage , and the ballistic shield which can be used to slam people who are close and blocks 66% of damage whilst deployed
Both of these shields only block FRONTAL attacks
The ballistic slam can be dodged by anyone with 30+ VIG

All of these are available either as maintenance spawns or moebius research

Misc additions
Added a extra examine text denoting that you can remove pilots by harm-intent clicking a mech.
Made full-auto CH compatible with mechs
Trying to stomp someone in space will instead push them below.
Refactored ammunition cloning to be more sane and centralized
Fixed every mech equipment from self-attack on any click regardless of context

## Why It's Good For The Game
Content and cool mechs
## Testing
Tested it locally for every feature added , and they all worked as expected
Tested ammo transfers and cloning,  to make sure they were not broken by the refactor there

## Changelog
:cl:
add: Added 2 new mech weapons the "Bobby" shotgun with 12 shells and instant recocking and the "C35R" smg with increased accuracy
add: Added the mech energy shield generator and ballistic shield
add: Added the mech thruster
add: Added direction-based functionaltiy to mechs (hatch-attacking and shooting)
add: Added a new mech meelle blade , which can be hand-crafted
add: New moebius designs for all of the mech equipments added
add: Clamps will now auto-insert ore from a walked-onto turf if they have an orebox
add: Spawnable variants of the mech blade in maintenance (cardboard,  plasteel , osmium and mettalic hydrogen)
tweak: Made the plasmacutter penetrate 5 tiles of mineral turfs
tweak: The mech PK is now 400 full auto instead of 5-burst
code: Reworked ammo cloning for insertion out and outside of containers(it uses New(bullet) now)
code: Added in a ballistics framework for mechs
code: Adjusted full auto click handlers and the swing attack proc to work with mechs

/:cl:

### CODE REVIEW

File: code/modules/mechs/equipment/combat.dm [hunk 1]
```diff
@@ -129,31 +758,267 @@
...
+				last_toggle = world.time
+				on = FALSE
+				update_icon()
+				return damages
+
+	return damages
+
+/obj/item/mech_equipment/shield_generator/ballistic
+	name = "ballistic mech shield"
+	desc = "A large, bulky shield meant to protect hunkering mechs"
```
Comment (hyperioo) [inline code/modules/mechs/equipment/combat.dm:901]:
```suggestion
	desc = "A large, bulky shield meant to protect hunkering mechs."
```

File: code/modules/mechs/equipment/combat.dm [hunk 2]
```diff
@@ -129,31 +758,267 @@
...
+	visual_bluff.icon_state = "mech_shield_[get_hardpoint()]"
+
+/obj/item/mech_equipment/shield_generator/ballistic/attack_self(mob/user)
+	var/mob/living/exosuit/mech = loc
+	if(!istype(mech))
+		return
+	to_chat(user , SPAN_NOTICE("[on ? "Retracting" : "Deploying"] \the [src]..."))
+	if(do_after(user, 3 SECOND, src, FALSE))
+		on = !on
+		to_chat(user, "You [on ? "deploy" : "retract"] \the [src]")
```
Comment (hyperioo) [inline code/modules/mechs/equipment/combat.dm:975]:
```suggestion
		to_chat(user, "You [on ? "deploy" : "retract"] \the [src].")
```

File: code/modules/mechs/equipment/combat.dm [hunk 3]
```diff
@@ -129,31 +758,267 @@
...
+			target = get_turf(target)
+		for(var/mob/living/knocked in target.contents)
+			targets.Add(knocked)
+		do_attack_animation(target, TRUE)
+
+		for(var/mob/living/knockable in targets)
+			if(ishuman(knockable))
+				var/mob/living/carbon/human/targ = knockable
+				if(targ.stats.getStat(STAT_VIG) > STAT_LEVEL_EXPERT)
+					targ.visible_message(SPAN_DANGER("[targ] dodges the shield slam!"), "You dodge [loc]'s shield slam!", "You hear a woosh", 6)
```
Comment (hyperioo) [inline code/modules/mechs/equipment/combat.dm:1002]:
```suggestion
					targ.visible_message(SPAN_DANGER("[targ] dodges the shield slam!"), "You dodge [loc]'s shield slam!", "You hear a woosh.", 6)
```

File: code/modules/mechs/mech.dm [hunk 4]
```diff
@@ -194,6 +196,8 @@
 		to_chat(user, "Its [thing.name] [thing.gender == PLURAL ? "are" : "is"] [damage_string].")
 
 	to_chat(user, "It menaces with reinforcements of [material].")
+	to_chat(user, SPAN_NOTICE("You can remove people inside by HARM intent clicking with your hand . The hatch must be opened."))
```
Comment (hyperioo) [inline code/modules/mechs/mech.dm:199]:
```suggestion
	to_chat(user, SPAN_NOTICE("You can remove people inside by HARM intent clicking with your hand. The hatch must be opened."))
```

File: code/modules/mechs/mech_damage.dm [hunk 5]
```diff
@@ -18,24 +18,56 @@
...
+// not enough made it in
+	if(damages[BRUTE] < round(I.force / 2))
+		visible_message("\The [src]'s shields block the blow!", 1, 2 ,5)
+		return
 
-	if(LAZYLEN(pilots) && !prob(body.pilot_coverage))
+	if(LAZYLEN(pilots) && ((!hatch_closed && (get_dir(user,src) & reverse_dir[dir])) || roll))
 		var/mob/living/pilot = pick(pilots)
+		var/turf/location = get_turf(src)
+		location.visible_message(SPAN_DANGER("\The [user] attacks the pilot inside of \the [src]"),1,5)
```
Comment (hyperioo) [inline code/modules/mechs/mech_damage.dm:65]:
```suggestion
		location.visible_message(SPAN_DANGER("\The [user] attacks the pilot inside of \the [src]."),1,5)
```

File: code/modules/mechs/mech_damage.dm [hunk 6]
```diff
@@ -140,3 +184,31 @@
...
+			adjustBruteLoss(split, body)
+		blocked++
+	if(legs)
+		adjustBruteLoss(split, legs)
+		blocked++
+	if(arms)
+		adjustBruteLoss(split, arms)
+		blocked++
+	if(damage > 400)
+		occupant_message("You feel the shockwave of a external explosion pass through your body!")
```
Comment (hyperioo) [inline code/modules/mechs/mech_damage.dm:211]:
```suggestion
		occupant_message("You feel the shockwave of an external explosion pass through your body!")
```

File: code/modules/mechs/mech_movement.dm [hunk 7]
```diff
@@ -40,6 +88,11 @@
 	var/mob/living/exosuit/exosuit = host
 	if(world.time < next_move)
 		return MOVEMENT_STOP
+	/// Added to handle with the case of ballistic shields (and probably other ones in the future.)
+	var/text = exosuit.moveBlocked()
+	if(length(text))
+		to_chat(mover, SPAN_NOTICE(text))
```
Comment (TheShown911) [inline code/modules/mechs/mech_movement.dm:94]: This code is crazy

File: code/modules/mechs/mech_interaction.dm [hunk 8]
```diff
@@ -276,9 +266,39 @@
...
+			chosen = loadable_guns[loadable_guns[1]]
+		switch(chosen.loadMagazine(I,user))
+			if(-1)
+				to_chat(user, SPAN_NOTICE("\The [chosen] does not accept this type of magazine."))
+			if(0)
+				to_chat(user, SPAN_NOTICE("\The [chosen] has no slots left in its ammunition storage"))
+			if(1)
+				to_chat(user, SPAN_NOTICE("You load \the [I] into \the [chosen]"))
+			if(2)
+				to_chat(user, SPAN_NOTICE("You partially reload one of the existing ammo magazines inside of \the [chosen]"))
```
Comment (hyperioo) [inline code/modules/mechs/mech_interaction.dm:292]:
```suggestion
				to_chat(user, SPAN_NOTICE("\The [chosen] has no slots left in its ammunition storage."))
			if(1)
				to_chat(user, SPAN_NOTICE("You load \the [I] into \the [chosen]."))
			if(2)
				to_chat(user, SPAN_NOTICE("You partially reload one of the existing ammo magazines inside of \the [chosen]."))
```

File: code/modules/mechs/mech_interaction.dm [hunk 9]
```diff
@@ -467,6 +487,10 @@
 	if(hatch_locked)
 		to_chat(user, SPAN_WARNING("The [body.hatch_descriptor] is locked."))
 		return
+	if(body && body.total_damage >= body.max_damage)
+		to_chat(user, SPAN_NOTICE("The chest of \the [src] is far too damaged.The hatch hinges are stuck!"))
```
Comment (hyperioo) [inline code/modules/mechs/mech_interaction.dm:491]:
```suggestion
		to_chat(user, SPAN_NOTICE("The chest of \the [src] is far too damaged. The hatch hinges are stuck!"))
```

File: code/modules/mechs/mech_movement.dm [hunk 10]
```diff
@@ -6,9 +6,57 @@
...
+			for(var/mob/living/victim in get_turf(src))
+				if(victim.lying || victim.resting || victim.mob_size <= MOB_MEDIUM)
+					if(blocked || victim.anchored)
+						victim.apply_damage(legs.stomp_damage, BRUTE, BP_CHEST, 1, 1.5, FALSE, FALSE, src.legs )
+						visible_message("The [src] stomps on [victim], crushing their chest!")
+						occupant_message("You can feel the [src] stomp something")
+					else
+						victim.forceMove(theDepths)
+						visible_message("The [src] pushes [victim] downwards")
+						occupant_message("You can feel the [src] step onto something")
```
Comment (hyperioo) [inline code/modules/mechs/mech_movement.dm:42]:
```suggestion
						occupant_message("You can feel \the [src] stomp something.")
					else
						victim.forceMove(theDepths)
						visible_message("The [src] pushes [victim] downwards.")
						occupant_message("You can feel \the [src] step onto something.")
```

File: code/modules/multiz/movement.dm [hunk 11]
```diff
@@ -56,13 +62,22 @@
 
 	//After checking that there's a valid destination, we'll first attempt phase movement as a shortcut.
 	//Since it can pass through obstacles, we'll do this before checking whether anything is blocking us
-	if(src.current_vertical_travel_method)
+	if(istype(mover, /mob/living/exosuit))
+		var/mob/living/mech = mover
+		if(mech.current_vertical_travel_method)
+			to_chat(src, SPAN_NOTICE("You can't do this yet"))
```
Comment (hyperioo) [inline code/modules/multiz/movement.dm:68]:
```suggestion
			to_chat(src, SPAN_NOTICE("You can't do this yet!"))
```

**Review by hyperioo — APPROVED** [review]

**Comment by hyperioo** [comment]
Try to be careful about the dots and stuff, it's a pain to make the suggestions and merge them through the web interface.

File: code/datums/datum_click_handlers.dm [hunk 12]
```diff
@@ -93,7 +93,7 @@
 	reciever.afterattack(target, owner.mob, FALSE)
 
 /datum/click_handler/fullauto/MouseDown(object, location, control, params)
-	if(!isturf(owner.mob.loc)) // This stops from firing full auto weapons inside closets or in /obj/effect/dummy/chameleon chameleon projector
+	if(!isturf(owner.mob.loc) && !ismech(owner.mob.loc)) // This stops from firing full auto weapons inside closets or in /obj/effect/dummy/chameleon chameleon projector
```
Comment (hyperioo) [inline code/datums/datum_click_handlers.dm:96]:
```suggestion
	if(!isturf(owner.mob.loc) && !ismech(owner.mob.loc)) // This stops from firing full auto weapons inside closets, in /obj/effect/dummy/chameleon chameleon projector or in a mech
```

**Pull request merged**